nengo/networks/linear_system.py

```diff
@@ -0,0 +1,216 @@
+    not perfect. If the input derivatives are known, then the accuracy can be
+    made perfect again. See references for details.
+
+    References
+    ----------
+    .. [#] A. R. Voelker and C. Eliasmith (2018) "Improving spiking dynamical networks:
+       Accurate delays, higher-order synapses, and time cells." Neural Computation.
+       [`URL <https://github.com/arvoelke/delay2017>`__]
+    """
+
+    if dt is not None and sys.analog:
+        sys = sys.discretize(dt=dt)
+
+    if dt is not None and synapse.analog:
+        synapse = synapse.discretize(dt=dt)
+
+    if synapse.analog != sys.analog:
+        raise ValueError("If `dt is None`, `sys.analog` must equal `synapse.analog`")
+
+    num, den = synapse.tf
+    if num.ndim == 2:
+        if num.shape[0] == 1:
+            num = num[0]
+        else:
+            raise NotImplementedError("Synapse must be a single-output system")
+
+    assert num.ndim == 1
+
+    num_max = np.max(num)
+    num[np.abs(num) < 1e-8 * num_max] = 0
+
+    num, den = np.poly1d(num), np.poly1d(den)
+    if synapse.analog and len(num.coeffs) > 1:
+        raise ValueError("analog synapse (%s) must not have zeros" % synapse)
+
+    # If the synapse was discretized, then its numerator may now have multiple
+    #   coefficients. By summing them together, we are implicitly assuming that
+    #   the output of the synapse will stay constant across
+    #   synapse.order_num + 1 time-steps. This is also related to:
+    #       http://dsp.stackexchange.com/questions/33510
+    # For example, if we have H = Lowpass(0.1), then the only difference
+    #   between sys1 = cont2discrete(H*H, dt) and
+    #           sys2 = cont2discrete(H, dt)*cont2discrete(H, dt), is that
+    #   np.sum(sys1.num) == sys2.num (while sys1.den == sys2.den)!
+    gain = np.sum(num)
+    c = den / gain
+
+    A, B, C, D = sys.ss
+    k = synapse.state_size
+    powA = [np.linalg.matrix_power(A, i) for i in range(k + 1)]
+
+    AH = np.sum([c[i] * powA[i] for i in range(k + 1)], axis=0)
+    BH = np.dot(
+        sum(c[i] * powA[i - j - 1] for j in range(k) for i in range(j + 1, k + 1)), B,
+    )
+
+    return LinearSystem((AH, BH, C, D), analog=sys.analog, x0=sys.x0)
+
+
+class LinearSystemNetwork(nengo.Network):
+    """Implement an arbitrary linear system in neurons.
+
+    Takes a linear system and uses `.ss2sim` to account for the effect of the
+    input/recurrent synapse on the system dynamics. The resulting network will provide
+    an accurate implementation of the system using the provided synapse.
+
+    Parameters
+    ----------
+    system : `.LinearSystem`
+        The LinearSystem to be implemented.
+    synapse : `.LinearFilter`
+        The synapse to use on the recurrent connection and input connection. Must be
+        an instance or subclass of `.LinearFilter`.
+    n_neurons : int
+        The number of neurons per ensemble in the state `.EnsembleArray`.
+    dt : float (optional)
```

Comment:
Think `, optional` is used everywhere instead of `(optional)`. Unsure if it matters.